Add timestamps to Preservation tasks

Diogenesoftoronto opened the pull request:
This PR aims to add timestamps to Preservations Tasks and give additional detail to the user in the enduro dashboard about when archivematica jobs started and finished.

Closes #832

Diogenesoftoronto committed:
Add timestamps to Preservation tasks

Adds timestamps to Preservations Tasks and gives additional details
to the user in the enduro dashboard about when archivematica jobs
started and finished.

Closes #832

sevein commented on internal/am/job_tracker.go:
```diff
@@ -115,9 +117,36 @@ func filterSavedJobs(jobs []amclient.Job, saved map[string]struct{}) []amclient.
 }
+
+func getJobTimestamp(job amclient.Job) (sql.NullTime, sql.NullTime) {
+	st := job.Tasks[0].StartedAt.Time
+	ct := job.Tasks[0].CompletedAt.Time
```
You're assuming that the slice of tasks isn't empty but that could not always be true, e.g. a bug in the API server. We must check the length of the slice before to avoid a potential panic (index out of range).

sevein commented on internal/am/job_tracker.go:
```diff
@@ -115,9 +117,36 @@ func filterSavedJobs(jobs []amclient.Job, saved map[string]struct{}) []amclient.
 	}
 }
+
+func getJobTimestamp(job amclient.Job) (sql.NullTime, sql.NullTime) {
```
The name `getJobTimestamp` can be misleading because it implies that the function returns a single timestamp. A better name could be `jobTimeRange`. I've omitted the "get" prefix because of this recommendation: https://go.dev/doc/effective_go#Getters.

sevein commented on internal/am/job_tracker.go:
```diff
@@ -115,9 +117,36 @@ func filterSavedJobs(jobs []amclient.Job, saved map[string]struct{}) []amclient.
+	st := job.Tasks[0].StartedAt.Time
+	ct := job.Tasks[0].CompletedAt.Time
+
+	for _, t := range job.Tasks {
```
Since [0] is already assigned, I think here you could use `range job.Tasks[1:]`, right?

sevein replied:
Since [0] is already assigned, I think here you could use `range job.Tasks[1:]`, right?

sevein commented on internal/am/job_tracker.go:
```diff
@@ -115,9 +117,36 @@ func filterSavedJobs(jobs []amclient.Job, saved map[string]struct{}) []amclient.
+	ct := job.Tasks[0].CompletedAt.Time
+
+	for _, t := range job.Tasks {
+		// if current st is after the task time set that as the startedat time.
```
Go doc comments use complete sentences, also meaning that they start with capital letter.

```suggestion
		// If current st is after the task time set that as the startedat time.
```

sevein reviewed (changes requested):
I've added a few comments. Have you confirmed that the timestamps are being properly recorded? In the `savePreservationTasks` method, I see the following:

```go
		pt := ConvertJobToPreservationTask(job)
		pt.PreservationActionID = jt.presActionID

		now := sql.NullTime{Time: jt.clock.Now(), Valid: true}
		pt.StartedAt = now
		pt.CompletedAt = now
```

With your changes, ConvertJobToPreservationTask is populating the timestamps but they're set to `now` right after. If you tested this PR and observed the timestamps recorded I suppose that'd become apparent.

Diogenesoftoronto commented:
> I've added a few comments. Have you confirmed that the timestamps are being properly recorded? In the `savePreservationTasks` method, I see the following:
> 
> ```go
> 		pt := ConvertJobToPreservationTask(job)
> 		pt.PreservationActionID = jt.presActionID
> 
> 		now := sql.NullTime{Time: jt.clock.Now(), Valid: true}
> 		pt.StartedAt = now
> 		pt.CompletedAt = now
> ```
> 
> With your changes, ConvertJobToPreservationTask is populating the timestamps but they're set to `now` right after. If you tested this PR and observed the timestamps recorded I suppose that'd become apparent.

Only thing I am worried about for this is that the tasks could very well be completed before this current time such that if any task that would have had a different completedAt time would have `now` as their time. I think this is a problem. Too fix this I could compare to check if the first value in the list is actually after the time completed. If it is then we know the value is both not null and the time that the task actually completed.

Diogenesoftoronto commented:
> > I've added a few comments. Have you confirmed that the timestamps are being properly recorded? In the `savePreservationTasks` method, I see the following:
> > ```go
> > 		pt := ConvertJobToPreservationTask(job)
> > 		pt.PreservationActionID = jt.presActionID
> > 
> > 		now := sql.NullTime{Time: jt.clock.Now(), Valid: true}
> > 		pt.StartedAt = now
> > 		pt.CompletedAt = now
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > With your changes, ConvertJobToPreservationTask is populating the timestamps but they're set to `now` right after. If you tested this PR and observed the timestamps recorded I suppose that'd become apparent.
> 
> Only thing I am worried about for this is that the tasks could very well be completed before this current time such that if any task that would have had a different completedAt time would have `now` as their time. I think this is a problem. Too fix this I could compare to check if the first value in the list is actually after the time completed. If it is then we know the value is both not null and the time that the task actually completed.

I fixed this concern. I also added some of the tests. There is better test coverage and I individually tested both of the functions to make sure they return the correct times.

sevein commented on internal/am/job_tracker.go:
```diff
@@ -81,11 +84,16 @@ func (jt *JobTracker) savePreservationTasks(ctx context.Context, jobs []amclient
+		if !pt.CompletedAt.Valid {
+			pt.CompletedAt.Time = jt.clock.Now()
+			pt.CompletedAt.Valid = true
+		}
```
Should we get rid of this block entirely? The goal of this snippet seems to be avoid NULLs and use the current time instead. I guess my point is that we don't need this anymore since now we can retrieve timestamps from Archivematica.

sevein commented on internal/am/job_tracker.go:
```diff
@@ -115,9 +123,44 @@ func filterSavedJobs(jobs []amclient.Job, saved map[string]struct{}) []amclient.
+	end := sql.NullTime{
+		Time:  ct,
+		Valid: ct != t,
+	}
```
Perhaps this is a simpler alternative that should be functionally equivalent:

```go
	start := sql.NullTime{Time: st, Valid: !st.IsZero()}
	end := sql.NullTime{Time: ct, Valid: !ct.IsZero()}
```

sevein commented on internal/am/job_tracker_test.go:
```diff
@@ -161,3 +173,91 @@ func TestJobTracker(t *testing.T) {
+	type test struct {
+		name string
+		job  amclient.Job
+		want []sql.NullTime
```
Because this test is named TestConvertJobToPreservationTask I think it should test its return value (`package_.PreservationTask`) as opposed to part of it (timestamp fields).

sevein reviewed (changes requested):
I think this is almost complete but I've included additional comments. I also submitted 9a77effb824afffc1f03b06e36ec440c54a46925 myself to save you the extra work. Please let me know your thoughts.

Diogenesoftoronto commented:
It looks great @sevein, I didn't realize you were suggesting removing the null check here, but I suppose that makes sense. Thanks for implementing the solutions to the tests as well. They look great. I have rebased and formatted the commit. It is ready to merge.

sevein reviewed (approved):
:+1:
